[CALCITE-3111] Allow custom implementations of Correlate in RelDecorr…

jh3507 opened the pull request:
…elator

- Refactor all Logical rels into corresponding abstract rels
- Use given RelBuilder when creating a new rel
- Add abstract rel visitors in CorelMapBuilder
- Change access levels to allow extending RelDecorrelator
- Add a test case

Change-Id: Ieed2af07f5d4277a02fb8ec3d67c11801ce4098a

laurentgo commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -425,7 +435,10 @@ public Frame decorrelateRel(Sort rel) {
 
     final Sort newSort =
-        LogicalSort.create(newInput, newCollation, rel.offset, rel.fetch);
+        (Sort) relBuilder
```
is it guaranteed that relbuilder would return a sort (for example if newInput has already the correct collation?)

jh3507 replied:
I think it is. However, it seems like there is nothing to lose by using RelNode. Changed its type to RelNode.

laurentgo commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -388,7 +389,16 @@ public Frame decorrelateRel(RelNode rel) {
+   *
+   * @param rel Sort to be rewritten
+   */
+  public Frame decorrelateRel(LogicalSort rel) {
```
if this method didn't exist before, there's no need to add it for binary compatibility.

jh3507 replied:
Done.

laurentgo commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -442,11 +455,11 @@ public Frame decorrelateRel(Values rel) {
    * @param rel Aggregate to rewrite
    */
-  public Frame decorrelateRel(LogicalAggregate rel) {
+  public Frame decorrelateRel(Aggregate rel) {
```
Here, we would need to keep `decorrelateRel(LogicalAggregate)` for binary compatibility...

jh3507 replied:
Done.

laurentgo commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -1219,10 +1283,12 @@ public Frame decorrelateRel(LogicalJoin rel) {
+        .push(leftFrame.r)
+        .push(rightFrame.r)
+        .join(rel.getJoinType(), decorrelateExpr(currentRel, map, cm, rel.getCondition()),
+            ImmutableSet.<CorrelationId>of())
```
Couldn't the type be inferred?

jh3507 replied:
Yes. I removed all <CorrelationId> from my change.

laurentgo commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -2310,9 +2376,9 @@ public void onMatch(RelOptRuleCall call) {
-          LogicalJoin.create(left, right, joinCond,
-              ImmutableSet.of(), joinType);
+      Join join =
+          (Join) relBuilder.push(left).push(right)
```
is it guaranteed to return a join?

jh3507 replied:
We are not calling the function with correlated fields, and, if that is the case, this join builder always seems to call createJoin method from factory. Alternatively, I think we could check the type of it and only call projectJoinOutputWithNullability function if it is join instance.

laurentgo commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -2656,10 +2718,10 @@ public int compareTo(@Nonnull CorDef o) {
-    private final Map<RexFieldAccess, CorRef> mapFieldAccessToCorRef;
+  protected static class CorelMap {
+    protected final Multimap<RelNode, CorRef> mapRefRelToCorRef;
+    public final SortedMap<CorrelationId, RelNode> mapCorToCorRel;
```
this seems "dangerous" to make the set public

jh3507 replied:
I changed all members to private and created a getter for mapCorToCorRel. Thanks for noticing this.

julianhyde commented on core/src/main/java/org/apache/calcite/tools/RelBuilder.java:
```diff
@@ -2286,6 +2301,13 @@ public RelBuilder sortLimit(int offset, int fetch, RexNode... nodes) {
+    replaceTop(sort);
+    return this;
+  }
+
```
This method is not necessary. Instead of `builder.sortLimit(collation, offset, fetch)` you can call `builder.sortLimit(offset, fetch, builder.fields(collation))`.

jh3507 replied:
Done. Thanks for the suggestion.

julianhyde commented on core/src/main/java/org/apache/calcite/tools/RelBuilder.java:
```diff
@@ -1978,6 +1978,21 @@ public RelBuilder join(JoinRelType joinType, RexNode condition,
+    stack.push(new Frame(join, fields.build()));
+    return this;
+  }
+
```
To be consistent with other methods in `RelBuilder` I don't think `requiredColumns` would be an `ImmutableBitSet`. It would be something a bit friendlier. Write some tests in RelBuilderTest and see which API is most user-friendly.

jh3507 replied:
I changed it to a list of integers(using varargs could be another option) and added a test case. I think this is a bit more user-friendly.

julianhyde replied:
The test you added is ```
RelNode root = builder.scan("EMP")
        .variable(v)
        .scan("DEPT")
        .filter(
            builder.equals(builder.field(0),
                builder.field(v.get(), "DEPTNO")))
        .correlate(JoinRelType.LEFT, v.get().id,
            Lists.newArrayList(
                SqlTypeUtil.findField(builder.peek(1).getRowType(), "DEPTNO")))
        .build();
```

I think that the test should be
```
RelNode root = builder.scan("EMP")
        .variable(v)
        .scan("DEPT")
        .filter(
            builder.equals(builder.field(0),
                builder.field(v.get(), "DEPTNO")))
        .correlate(JoinRelType.LEFT, v.get().id, builder.field(2, 0, "DEPTNO"))
        .build();
```

That's to say, the signature should be
```
public RelBuilder correlate(JoinRelType joinType,
      CorrelationId correlationId, RexNode... requiredFields)
```
and also, the overloaded method
```
public RelBuilder correlate(JoinRelType joinType,
      CorrelationId correlationId, Iterable<? extends RexNode> requiredFields)
```

Normally, as in this case, `requiredFields` is one `RexInputRef`. But it might be several expressions, and if so, `correlate` should create an extra `Project`. (See how `aggregate(GroupKey, Iterable<AggCall>)` uses a `Registrar` to project extra expressions.)

jh3507 replied:
Thanks for the suggestion. I updated signature. So, in correlate case, I think we should create an extra project on right relation if given fields are not simply RexInputRefs. Added a new test case for that as well.

jh3507 commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -787,10 +812,29 @@ private RelNode createValueGenerator(
+            return fieldNames.get(pos);
+          }
+        };
+        RelNode distinct = relBuilder.push(newInput).projectNamed(exprs, names, false)
```
Done. That would make it much simpler.

julianhyde commented:
I'm "meh" on the javadoc for all the `decorrelateRel` methods. The methods are called via reflective dispatch so they all do the same thing. Adding javadoc doesn't add any information content.

Don't put ' );' on a new line. We don't format code that way.

Please add commits, don't amend or squash. It's difficult to review when you squash each time. (Some reviewers ask for squash and rebase. I guess I disagree with those reviewers. I'm happy to squash and rebase when I merge your PR.)

I think you need to change algebra.md. 

I hate code formatted like this:

```
        RelNode distinct = relBuilder.push(newInput).projectNamed(
            positions.stream().map(i -> relBuilder.getRexBuilder()
                .makeInputRef(newInput, i)).collect(Collectors.toList()),
            positions.stream().map(fieldNames::get).collect(Collectors.toList()),
            false)
             .distinct()
             .build();
```

Breaking the line at the line length limit seems to me like passive-aggressive adherence to the line length policy. Isn't this more readable?

```
        RelNode distinct = relBuilder.push(newInput)
            .projectNamed(positions.stream()
                    .map(i ->
                        relBuilder.getRexBuilder().makeInputRef(newInput, i))
                    .collect(Collectors.toList()),
                positions.stream().map(fieldNames::get).collect(Collectors.toList()),
                false)
            .distinct()
            .build();
```

That said, wouldn't 
```
        RelNode distinct = relBuilder.push(newInput)
            .project(relBuilder.fields(positions))
            .distinct()
            .build();
```
do the same thing?

jh3507 commented:
I agree that Javadoc on each decorrelateRel function doesn't give much information(I guess this would be also true for trimFields functions). I updated javadoc and fixed some code style. Thanks.
